src/lib/core/utils/data-element-constraints.ts

```diff
@@ -0,0 +1,123 @@
+  );
+}
+
+export type ValueByInputName = Partial<Record<string, Variable>>;
+export type DataElementConstraintRecord = Record<string, DataElementConstraint>;
+
+/**
+ * Given an array of DataElementConstraint objects and a set of values, return
+ * a unioned DataElementConstraint object that includes all of the rules for
+ * which the provided values satisfy.
+ */
+export function flattenConstraints(
+  values: ValueByInputName,
+  entities: StudyEntity[],
+  constraints: DataElementConstraintRecord[]
+): DataElementConstraintRecord {
+  // Find all compatible constraints
+  const compatibleConstraints = constraints.filter((constraintRecord) =>
+    Object.entries(constraintRecord).every(([variableName, constraint]) => {
+      const value = values[variableName];
+      // If a value has not been selected for this constraint, then it is considered to be "in-play"
+      if (value == null) return true;
+      // If a constraint does not declare shapes or types, then any value is allowed, thus the constraint is "in-play"
+      if (isEmpty(constraint.allowedShapes) || isEmpty(constraint.allowedTypes))
+        return true;
+      // Check that the value's associated variable has a compatible type and shape
+      const entityAndVariable = findEntityAndVariable(entities, value);
+      if (entityAndVariable == null)
+        throw new Error(
+          `Could not find selected entity and variable: entityId = ${value.entityId}; variableId = ${value.variableId}.`
+        );
+      const { variable } = entityAndVariable;
+      const typeIsValid =
+        isEmpty(constraint.allowedTypes) ||
+        constraint.allowedTypes?.includes(variable.type);
+      const shapeIsValid =
+        isEmpty(constraint.allowedShapes) ||
+        constraint.allowedShapes?.includes(variable.dataShape!);
+      return typeIsValid && shapeIsValid;
+    })
+  );
+  if (compatibleConstraints.length === 0)
+    throw new Error(
+      'flattenConstraints: Something went wrong. No compatible constraints were found for the current set of values.'
+    );
+  // Combine compatible constraints into a single constraint, concatenating
+  // allowed shapes and types.
+  return compatibleConstraints.reduce(
+    mergeConstraints,
+    {} as DataElementConstraintRecord
+  );
+}
+
+export function mergeConstraints(
+  constraintMapA: DataElementConstraintRecord,
+  constraintMapB: DataElementConstraintRecord
+): DataElementConstraintRecord {
+  const keys = union(Object.keys(constraintMapA), Object.keys(constraintMapB));
+  return Object.fromEntries(
+    keys.map((key): [string, DataElementConstraint] => {
+      const constraintA = constraintMapA[key];
+      const constraintB = constraintMapB[key];
+      return [
+        key,
+        constraintA == null
+          ? constraintB
+          : constraintB == null
+          ? constraintA
+          : {
+              isRequired: constraintA.isRequired || constraintB.isRequired,
+              maxNumVars: Math.max(
```

Comment:
My hunch says the maxNumVars should be merged with Math.min() (and likewise min with max).

Merging two constraints that are min: 1, max: 6, and min: 2 and max: 3 should merge to min: 2, max: 3, right?

If the two constrains were min:1, max:2 and min:4, max:5, then these couldn't be merged, right?

Also, how are multiple (e.g. 2x facet) variables handled by InputVariables?  Can `InputVariables.props.values` store two variables for `facetVariable`?

Reply:
I think you are right. This is incomplete work. I think it can be addressed in a follow-up.

Reply:
I fixed the min/max issue. I'm not sure how to handle incompatible cases. It might put a damper on this entire approach!

Reply:
It may be safe to assume that these values will not vary across constraints. It's hard to imagine a case where {min,max}NumVars would change based on other criteria. We could start with throwing an error if the numbers aren't the same.

Reply:
Yeah I agree - it's unlikely that e.g. facetVariable can take 5-7 variables for "date vs number" plots and 1-3 variables for "number vs number".

Definitely a can to kick down the road if we put in place the safeguard you mention.

> Also, how are multiple (e.g. 2x facet) variables handled by InputVariables? Can InputVariables.props.values store two variables for facetVariable?

Is this also something to deal with when we have a faceting strategy (on the components side of things) ?

Reply:
Yes, I think we can deal with it then.